### PR TITLE
GATEWAY - 3028

### PR DESCRIPTION
**Pull request by bhumphrey**
Comprises of a series of refactors to clean up the code and improve the readability. It is by far from complete as there is still massive amounts of duplicated code. 
-  pulled common elements up to higher in class hierarchy
- extracted methods for common functionality
- introduced interfaces to decouple implementation

### CODE REVIEW

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    /**
+    
+    /**
+     * This method will return a list of all business entities that are known by the connection manager.
+     * 
+     * @return The list of all business entities known by the connection manager.
+     * @throws ConnectionManagerException
+     */
+    public List<BusinessEntity> getAllBusinessEntities() throws ConnectionManagerException;
```
Comment (tparnell) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java:50]: List seems like the wrong semantic here.  Are they ordered?  Are repeats allowed?  Seems like Set is better (and aligns with getBusinessEntitySet API).

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+     * regardless of whether it is in the UDDI cache or not.
+     * 
+     * @param sUniformServiceName The name of the service being searched for.
+     * @return The business entities that have this service defined.
+     * @throws gov.hhs.fha.nhinc.connectmgr.ConnectionManagerException
+     */
+    public Set<BusinessEntity> getAllBusinessEntitySetByServiceName(String sUniformServiceName)
+            throws ConnectionManagerException;
+
+    public List<UDDI_SPEC_VERSION> getSpecVersions(String homeCommunityId, NhincConstants.NHIN_SERVICE_NAMES serviceName);
```
Comment (tparnell) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java:124]: Same comment re:List vs. Set.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java (hunk 1 above)
Comment (bhumphrey) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java:50]: totally agree. I just extracted an interface - reusing the existing the methods

Comment (bhumphrey) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/connectmgr/ConnectionManager.java:50]: created improvement ticket: https://issues.connectopensource.org/browse/GATEWAY-3172

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java [hunk 3]
```diff
@@ -44,86 +50,164 @@
...
+
+        auditRequestMessage(message.getRequest(), message.getAssertion(), message.getTarget().getHomeCommunity()
+                .getHomeCommunityId());
+        try {
+            executeStrategy(message);
+        } catch (Exception ex) {
+            handleError(message, ex);
+        }
+
+        auditResponseMessage(message.getResponse(), message.getAssertion(), message.getTarget().getHomeCommunity()
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java:123]: This looks wrong.  I don't think you can audit after handleError - there's no guarantee that message.getResponse() is anything sane, is there?  Or does auditResponseMessage handle null & other non-sensical values?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java [hunk 4]
```diff
@@ -44,86 +50,164 @@
...
+     * @throws ConnectionManagerException
+     * @throws IllegalArgumentException
+     */
+    @SuppressWarnings("static-access")
+    public void executeStrategy(OutboundDocQueryOrchestratable message) throws IllegalArgumentException,
+            ConnectionManagerException, Exception {
+        
+        final String url = getUrl(message.getTarget());
+        message.getTarget().setUrl(url)
+        ;
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java]: Formatter.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java [hunk 5]
```diff
@@ -44,86 +50,164 @@
...
+        if (log.isDebugEnabled()) {
+            log.debug("executeStrategy sending nhin doc query request to " + " target hcid="
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategy.java:141]: style preference: I like to use String.format() for these types of things - easier to read.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategyImpl_g1.java [hunk 6]
```diff
@@ -39,79 +34,36 @@
...
-            message.setResponse(((OutboundDocQueryOrchestratable_a1) processor.processErrorResponse(message, err))
-                    .getResponse());
-            getLogger().debug("NhinDocQueryStrategyImpl_g1::executeStrategy returning error response");
-        }
-        auditResponseMessage(message.getResponse(), NhincConstants.AUDIT_LOG_INBOUND_DIRECTION,
-                NhincConstants.AUDIT_LOG_NHIN_INTERFACE, message.getAssertion(), message.getTarget().getHomeCommunity()
-                        .getHomeCommunityId());
+    @Override
+    protected GATEWAY_API_LEVEL getAPILevel() {
+        // TODO Auto-generated method stub
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategyImpl_g1.java]: punt this line.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/nhin/proxy/NhinDocQueryProxyFactory.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package gov.hhs.fha.nhinc.docquery.nhin.proxy;
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/nhin/proxy/NhinDocQueryProxyFactory.java:27]: Missing file header.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 8]
```diff
@@ -154,32 +158,45 @@ public AdhocQueryResponse respondingGatewayCrossGatewayQuery(AdhocQueryRequest a
...
 
         try {
-            if (targets != null && NullChecker.isNotNullish(targets.getNhinTargetCommunity())) {
-                isTargeted = true;
-            }
             // Obtain all the URLs for the targets being sent to
             urlInfoList = getEndpointForNhinTargetCommunities(targets, NhincConstants.DOC_QUERY_SERVICE_NAME);
 
-            if ((!isTargeted) || (isTargeted && NullChecker.isNotNullish(urlInfoList))) {
+            if (isTargeted(targets)) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java]: why not isTargeted |= isTargeted(targets);

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 9]
```diff
@@ -197,44 +214,23 @@ public AdhocQueryResponse respondingGatewayCrossGatewayQuery(AdhocQueryRequest a
...
-                                    }
-                                }
+                                setPatientIdOnRequest(clonedRequest,
+                                        identifier.getSubjectIdentifier(), identifier.getAssigningAuthorityIdentifier());
                                 // set the home community id to the target hcid
-                                clonedRequest = setTargetHomeCommunityId(clonedRequest, sTargetHomeCommunityId);
+                                String targetHomeCommunityId = target.getHomeCommunity().getHomeCommunityId();
+                                setTargetHomeCommunityId(clonedRequest, targetHomeCommunityId);
                                 for (UrlInfo urlInfo : urlInfoList) {
-                                    if (urlInfo.getHcid().equals(sTargetHomeCommunityId)) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java:237]: I don't understand this change.  It looks substantive.  Let's chat.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 10]
```diff
@@ -296,6 +295,38 @@ public AdhocQueryResponse respondingGatewayCrossGatewayQuery(AdhocQueryRequest a
...
+     */
+    protected String setPatientIdOnRequest(AdhocQueryRequest request, String subjectIdentifier,
+            String assigningAuthorityIdentifier) {
+        String formattedPatientId = PatientIdFormatUtil.hl7EncodePatientId(subjectIdentifier,
+                assigningAuthorityIdentifier);
+        List<SlotType1> slotType1 = request.getAdhocQuery().getSlot();
+        Iterator<SlotType1> iterSlotType1 = slotType1.iterator();
+        while (iterSlotType1.hasNext()) {
+            SlotType1 slot = iterSlotType1.next();
+            if ((slot.getName() != null)
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java]: Simplification:
     if (DocumentTransformConstants.EBXML_DOCENTRY_PATIENT_ID.equals(slot.getName()) {
handles the null case without extra logic.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 11]
```diff
@@ -296,6 +295,38 @@ public AdhocQueryResponse respondingGatewayCrossGatewayQuery(AdhocQueryRequest a
...
+     * @param assigningAuthorityIdentifier
+     * @return
+     */
+    protected String setPatientIdOnRequest(AdhocQueryRequest request, String subjectIdentifier,
+            String assigningAuthorityIdentifier) {
+        String formattedPatientId = PatientIdFormatUtil.hl7EncodePatientId(subjectIdentifier,
+                assigningAuthorityIdentifier);
+        List<SlotType1> slotType1 = request.getAdhocQuery().getSlot();
+        Iterator<SlotType1> iterSlotType1 = slotType1.iterator();
+        while (iterSlotType1.hasNext()) {
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java]: Why not just
for (SlotType1 slot : request.getAdhocQuery().getSlot()) {
?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java [hunk 12]
```diff
@@ -315,19 +346,18 @@ private AdhocQueryResponse createPatientIdError(List<SlotType1> slotList, AdhocQ
...
      * @return
      */
-    private AdhocQueryRequest setTargetHomeCommunityId(AdhocQueryRequest clonedRequest, String sTargetHomeCommunityId) {
+    private AdhocQueryRequest setTargetHomeCommunityId(AdhocQueryRequest request, String sTargetHomeCommunityId) {
         if (NullChecker.isNotNullish(sTargetHomeCommunityId)) {
             if (!(sTargetHomeCommunityId.startsWith("urn:oid:"))) {
-                clonedRequest.getAdhocQuery().setHome("urn:oid:" + sTargetHomeCommunityId);
-            } else {
-                clonedRequest.getAdhocQuery().setHome(sTargetHomeCommunityId);
+                sTargetHomeCommunityId = "urn:oid:" + sTargetHomeCommunityId;
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java:351]: Why is this logic gone?  Do you know for sure that sTargetHomeCommunityId is unprefixed at this point?

File: Product/pom.xml [hunk 13]
```diff
@@ -318,6 +318,33 @@
                     <testTarget>${compiler.target}</testTarget>
                 </configuration>
             </plugin>
+ 			  <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
```
Comment (tparnell) [inline Product/pom.xml:322]: This is unrelated & probably should be merged independently.

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java (hunk 12 above)
Comment (bhumphrey) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java:351]: The prefix is only added when it's not there. 

```
if (NullChecker.isNotNullish(sTargetHomeCommunityId)) {
            if (!(sTargetHomeCommunityId.startsWith("urn:oid:"))) {
                sTargetHomeCommunityId = "urn:oid:" + sTargetHomeCommunityId;
            }
            request.getAdhocQuery().setHome(sTargetHomeCommunityId);
        }
        return request;
```

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java (hunk 9 above)
Comment (bhumphrey) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/outbound/StandardOutboundDocQuery.java:237]: i just changed the variable name form sTargetHomeCommunityId to targetHomeCommunityId

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategyImpl_g1.java [hunk 14]
```diff
@@ -39,79 +34,35 @@
...
-
-            message.setResponse(proxy.respondingGatewayCrossGatewayQuery(message.getRequest(), message.getAssertion(),
-                    message.getTarget()));
-
-            getLogger().debug("NhinDocQueryStrategyImpl_g1::executeStrategy returning response");
-        } catch (Exception ex) {
-            String err = ExecutorServiceHelper.getFormattedExceptionInfo(ex, message.getTarget(),
-                    message.getServiceName());
-            OutboundResponseProcessor processor = message.getResponseProcessor();
-            message.setResponse(((OutboundDocQueryOrchestratable_a1) processor.processErrorResponse(message, err))
```
Comment (tparnell) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryStrategyImpl_g1.java:109]: This logic needs to be pulled up.
